### PR TITLE
Connect AI Chat levels to continue button

### PR DESCRIPTION
**Pull request by sanchitmalhotra126**
Connect AI Chat levels to the continue button, which turns the progress bubble green. I had to rework some of the lab2 progress/instructions internals a bit and am curious how Lab2 folks feel about these logic updates, but I also expect that we'll continue to iterate on these pieces as we add more labs and explore functionality in AI Chat.

I also added a new AI Chat level to allthethings to test out moving between levels.

https://github.com/code-dot-org/code-dot-org/assets/85528507/1899f67d-cd04-4eca-be9a-49a153c868e6

## Links

https://codedotorg.atlassian.net/browse/LABS-25

## Testing story

Tested locally.

### CODE REVIEW

File: apps/src/lab2/views/components/Instructions.tsx [hunk 1]
```diff
@@ -47,11 +49,15 @@ const Instructions: React.FunctionComponent<InstructionsProps> = props => {
...
-  const validationState = useSelector(
+  const {hasConditions, message, satisfied} = useSelector(
     (state: {lab: LabState}) => state.lab.validationState
   );
+
+  // If there are no validation conditions, we can show the next button so long as
+  // there is another level. If validation is present, also check that conditions are satisfied.
   const showNextButton =
-    validationState.satisfied && levelIndex + 1 < currentLevelCount;
+    (!hasConditions || satisfied) && levelIndex + 1 < currentLevelCount;
```
Comment (sanchitmalhotra126) [inline apps/src/lab2/views/components/Instructions.tsx:64]: @molly-moen @breville want to get a quick gut check from y'all on this - once again trying to navigate that divide between keeping things simple for labs vs being too prescriptive and locking ourselves into patterns that we may have to undo later. We're making an assumption here that if a level has no validation conditions, then we are ok to show the continue button (so long as there is another level). On one hand this keeps things simple as neither AI Chat nor Music Lab has to worry about specifying the conditions under which to show the button, but this does introduce more conditional logic here.

Comment (breville) [reply apps/src/lab2/views/components/Instructions.tsx:64]: This seems reasonable, especially for now.

One nuance we might face later is that we probably want Music Lab's "Continue" button to appear a little bit later than when we hit the success condition, so that the user can see/hear the successful condition and then get the reinforcement of that button appearing.  It feels that similarly there might be some benefit in making the "Continue" button appear with a slight delay in these AI Chat levels, even if technically the user can skip immediately.  But maybe we should leave such delays up to the lab implementations.

**Review by breville — APPROVED** [review]

File: apps/src/lab2/progress/ProgressManager.ts [hunk 2]
```diff
@@ -32,16 +34,18 @@ export default class ProgressManager {
...
-    this.currentValidationState = initialValidationState;
+    this.currentValidationState = getInitialValidationState();
   }
 
   /**
    * Update the ProgressManager with level data for a new level.
    * Resets validation status internally.
    */
-  onLevelChange(levelData: ProjectLevelData) {
+  onLevelChange(levelData: ProjectLevelData | undefined) {
```
Comment (levadadenys) [inline apps/src/lab2/progress/ProgressManager.ts]: nit: can be simplified to `onLevelChange(levelData?: ProjectLevelData) {`

File: apps/src/lab2/views/components/Instructions.tsx [hunk 3]
```diff
@@ -64,9 +70,14 @@ const Instructions: React.FunctionComponent<InstructionsProps> = props => {
   return (
     <InstructionsPanel
       text={instructionsText}
-      message={validationState.message || undefined}
+      message={message || undefined}
       showNextButton={showNextButton}
-      onNextPanel={() => dispatch(navigateToNextLevel())}
+      onNextPanel={() => {
```
Comment (levadadenys) [inline apps/src/lab2/views/components/Instructions.tsx]: nit: what do you think abobut wrapping this with `useCallback`?

**Review by levadadenys — APPROVED** [review]
LGTM!
Left couple non blocking nit comments/suggestions

File: apps/src/lab2/views/components/Instructions.tsx (hunk 1 above)
Comment (sanchitmalhotra126) [reply apps/src/lab2/views/components/Instructions.tsx:64]: Yeah, that makes sense. I was thinking about if that delay technically qualifies as a validation condition ('user has to wait x seconds') and handling it through the progress system, or if each lab should just handle it in their own way. I'm sure we'll revisit this as we encounter more use cases.

Comment (molly-moen) [reply apps/src/lab2/views/components/Instructions.tsx:64]: I think so long as the lab can provide the validation function and set `satisfied` as it chooses, that still gives us enough flexibility per lab.

File: apps/src/lab2/views/components/Instructions.tsx (hunk 3 above)
Comment (sanchitmalhotra126) [reply apps/src/lab2/views/components/Instructions.tsx]: Done!

**Review by molly-moen — APPROVED** [review]

File: apps/src/lab2/views/components/Instructions.tsx [hunk 4]
```diff
@@ -47,15 +54,26 @@ const Instructions: React.FunctionComponent<InstructionsProps> = props => {
...
   const levelIndex = useSelector(currentLevelIndex);
   const currentLevelCount = useSelector(levelCount);
-  const validationState = useSelector(
+  const {hasConditions, message, satisfied} = useSelector(
     (state: {lab: LabState}) => state.lab.validationState
   );
+
+  // If there are no validation conditions, we can show the next button so long as
+  // there is another level. If validation is present, also check that conditions are satisfied.
   const showNextButton =
```
Comment (fisher-alice) [inline apps/src/lab2/views/components/Instructions.tsx:63]: Just curious - for the last level in a progression, will the bubble turn green for the user at some point? Or is validation needed?

**Review by fisher-alice — APPROVED** [review]
Looks great!

File: apps/src/lab2/lab2Redux.ts [hunk 5]
```diff
@@ -16,6 +16,7 @@ import {
   ProjectManagerStorageType,
   ProjectSources,
   LevelProperties,
+  ProjectLevelData,
```
Comment (fisher-alice) [inline apps/src/lab2/lab2Redux.ts]: Just curious - why did you add this? I don't see it being used in file.

File: apps/src/lab2/views/components/Instructions.tsx (hunk 4 above)
Comment (sanchitmalhotra126) [reply apps/src/lab2/views/components/Instructions.tsx:63]: Good question! This is actually a case we don't handle in Lab2 yet, so I'm not sure. Looks like there is a lot of code in legacy labs that handles the finish/continue buttons, so we'll need to see what makes sense to adapt into Lab2. Filed https://codedotorg.atlassian.net/browse/LABS-36 to investigate.

File: apps/src/lab2/lab2Redux.ts (hunk 5 above)
Comment (sanchitmalhotra126) [reply apps/src/lab2/lab2Redux.ts]: Good catch! Removed
